batt_smbus: Improvements for accessing more SBS data

ksschwabe opened the pull request:
Also fixes the the PEC calculation for writing. See
http://cache.freescale.com/files/32bit/doc/app_note/AN4471.pdf and
http://www.ti.com/lit/an/sloa132/sloa132.pdf or more
details on SMBus reading and writing including with and without PECs.

LorenzMeier commented:
Thanks!

LorenzMeier commented:
@julianoes Could you please bring any changes from https://github.com/3drobotics/PX4Firmware-solo to the smart battery driver and integrate them with these?

ksschwabe commented:
@LorenzMeier I'm going to add a few more things for reading out the Battery Manufacturer name, device serial etc. I'll merge that in later today.

LorenzMeier commented:
Ok, sounds great!

LorenzMeier commented:
@ksschwabe Could you have a look at this diff and see if / how you could integrate it?

```diff
diff --git a/src/drivers/batt_smbus/batt_smbus.cpp b/src/drivers/batt_smbus/batt_smbus.cpp
index 1b89655..42e4b83 100644
--- a/src/drivers/batt_smbus/batt_smbus.cpp
+++ b/src/drivers/batt_smbus/batt_smbus.cpp
@@ -39,7 +39,7 @@
  * @author Randy Mackay <rmackay9@yahoo.com>
  */
 
-#include <nuttx/config.h>
+#include <px4_config.h>
 
 #include <sys/types.h>
 #include <stdint.h>
@@ -77,25 +77,23 @@
 #include <drivers/drv_batt_smbus.h>
 #include <drivers/device/ringbuffer.h>
 
-#define BATT_SMBUS_ADDR_MIN				0x08	///< lowest possible address
-#define BATT_SMBUS_ADDR_MAX				0x7F	///< highest possible address
+#define BATT_SMBUS_ADDR_MIN			0x08	///< lowest possible address
+#define BATT_SMBUS_ADDR_MAX			0x7F	///< highest possible address
 
-#define BATT_SMBUS_I2C_BUS				PX4_I2C_BUS_EXPANSION
-#define BATT_SMBUS_ADDR					0x0B	///< I2C address
-#define BATT_SMBUS_TEMP					0x08	///< temperature register
-#define BATT_SMBUS_VOLTAGE				0x09	///< voltage register
+#define BATT_SMBUS_I2C_BUS			PX4_I2C_BUS_EXPANSION
+#define BATT_SMBUS_ADDR				0x0B	///< I2C address
+#define BATT_SMBUS_TEMP				0x08	///< temperature register
+#define BATT_SMBUS_VOLTAGE			0x09	///< voltage register
 #define BATT_SMBUS_REMAINING_CAPACITY	0x0f	///< predicted remaining battery capacity as a percentage
 #define BATT_SMBUS_FULL_CHARGE_CAPACITY 0x10    ///< capacity when fully charged
 #define BATT_SMBUS_DESIGN_CAPACITY		0x18	///< design capacity register
 #define BATT_SMBUS_DESIGN_VOLTAGE		0x19	///< design voltage register
 #define BATT_SMBUS_SERIALNUM			0x1c	///< serial number register
 #define BATT_SMBUS_MANUFACTURE_NAME		0x20	///< manufacturer name
-#define BATT_SMBUS_MANUFACTURE_DATA		0x23	///< manufacturer data
 #define BATT_SMBUS_MANUFACTURE_INFO		0x25	///< cell voltage register
-#define BATT_SMBUS_CURRENT				0x2a	///< current register
-#define BATT_SMBUS_MEASUREMENT_INTERVAL_US	(1000000 / 10)	///< time in microseconds, measure at 10hz
-#define BATT_SMBUS_TIMEOUT_US			10000000	///< timeout looking for battery 10seconds after startup
-#define BATT_SMBUS_BUTTON_DEBOUNCE_MS	300		///< button holds longer than this time will cause a power off event
+#define BATT_SMBUS_CURRENT			0x2a	///< current register
+#define BATT_SMBUS_MEASUREMENT_INTERVAL_MS	(1000000 / 10)	///< time in microseconds, measure at 10hz
+#define BATT_SMBUS_TIMEOUT_MS		10000000	///< timeout looking for battery 10seconds after startup
 
 #define BATT_SMBUS_PEC_POLYNOMIAL		0x07	///< Polynomial for calculating PEC
 
@@ -189,7 +187,6 @@ private:
 	orb_id_t		_batt_orb_id;	///< uORB battery topic ID
 	uint64_t		_start_time;	///< system time we first attempt to communicate with battery
 	uint16_t		_batt_capacity;	///< battery's design capacity in mAh (0 means unknown)
-	uint8_t			_button_press_counts; ///< count of button presses detected
 };
 
 namespace
@@ -202,15 +199,14 @@ void batt_smbus_usage();
 extern "C" __EXPORT int batt_smbus_main(int argc, char *argv[]);
 
 BATT_SMBUS::BATT_SMBUS(int bus, uint16_t batt_smbus_addr) :
-	I2C("batt_smbus", BATT_SMBUS0_DEVICE_PATH, bus, batt_smbus_addr, 100000),
+	I2C("batt_smbus", BATT_SMBUS_DEVICE_PATH, bus, batt_smbus_addr, 400000),
 	_enabled(false),
 	_work{},
 	_reports(nullptr),
-	_batt_topic(-1),
+	_batt_topic(nullptr),
 	_batt_orb_id(nullptr),
 	_start_time(0),
-	_batt_capacity(0),
-	_button_press_counts(0)
+	_batt_capacity(0)
 {
 	// work_cancel in the dtor will explode if we don't do this...
 	memset(&_work, 0, sizeof(_work));
@@ -301,8 +297,8 @@ BATT_SMBUS::test()
 
 		if (updated) {
 			if (orb_copy(ORB_ID(battery_status), sub, &status) == OK) {
-				warnx("V=%4.2f C=%4.2f DismAh=%4.2f Cap:%d Shutdown:%d", (float)status.voltage_v, (float)status.current_a,
-				      (float)status.discharged_mah, (int)_batt_capacity, (int)status.is_powering_off);
+				warnx("V=%4.2f C=%4.2f DismAh=%4.2f Cap:%d", (float)status.voltage_v, (float)status.current_a,
+				      (float)status.discharged_mah, (int)_batt_capacity);
 			}
 		}
 
@@ -385,7 +381,7 @@ BATT_SMBUS::cycle()
 	uint64_t now = hrt_absolute_time();
 
 	// exit without rescheduling if we have failed to find a battery after 10 seconds
-	if (!_enabled && (now - _start_time > BATT_SMBUS_TIMEOUT_US)) {
+	if (!_enabled && (now - _start_time > BATT_SMBUS_TIMEOUT_MS)) {
 		warnx("did not find smart battery");
 		return;
 	}
@@ -407,7 +403,7 @@ BATT_SMBUS::cycle()
 		new_report.voltage_v = ((float)tmp) / 1000.0f;
 
 		// read current
-		uint8_t buff[6];
+		uint8_t buff[4];
 
 		if (read_block(BATT_SMBUS_CURRENT, buff, 4, false) == 4) {
 			new_report.current_a = -(float)((int32_t)((uint32_t)buff[3] << 24 | (uint32_t)buff[2] << 16 | (uint32_t)buff[1] << 8 |
@@ -430,35 +426,14 @@ BATT_SMBUS::cycle()
 			}
 		}
 
-		// read the button press indicator
-		if (read_block(BATT_SMBUS_MANUFACTURE_DATA, buff, 6, false) == 6) {
-			bool pressed = (buff[1] >> 3) & 0x01;
-
-			if(_button_press_counts >= ((BATT_SMBUS_BUTTON_DEBOUNCE_MS * 1000) / BATT_SMBUS_MEASUREMENT_INTERVAL_US)) {
-				// battery will power off
-				new_report.is_powering_off = true;
-				// warn only once
-				if(_button_press_counts++ == ((BATT_SMBUS_BUTTON_DEBOUNCE_MS * 1000) / BATT_SMBUS_MEASUREMENT_INTERVAL_US)) {
-					warnx("system is shutting down NOW...");
-				}
-			} else if(pressed) {
-				// battery will power off if the button is held
-				_button_press_counts++;
-			} else {
-				// button released early, reset counters
-				_button_press_counts = 0;
-				new_report.is_powering_off = false;
-			}
-		}
-
 		// publish to orb
-		if (_batt_topic != -1) {
+		if (_batt_topic != nullptr) {
 			orb_publish(_batt_orb_id, _batt_topic, &new_report);
 
 		} else {
 			_batt_topic = orb_advertise(_batt_orb_id, &new_report);
 
-			if (_batt_topic < 0) {
+			if (_batt_topic == nullptr) {
 				errx(1, "ADVERT FAIL");
 			}
 		}
@@ -478,7 +453,7 @@ BATT_SMBUS::cycle()
 
 	// schedule a fresh cycle call when the measurement is done
 	work_queue(HPWORK, &_work, (worker_t)&BATT_SMBUS::cycle_trampoline, this,
-		   USEC2TICK(BATT_SMBUS_MEASUREMENT_INTERVAL_US));
+		   USEC2TICK(BATT_SMBUS_MEASUREMENT_INTERVAL_MS));
 }
 
 int
@@ -651,7 +626,7 @@ batt_smbus_main(int argc, char *argv[])
 		exit(0);
 	}
 
-	/* need the driver past this point */
+	// need the driver past this point
 	if (g_batt_smbus == nullptr) {
 		warnx("not started");
 		batt_smbus_usage();
diff --git a/src/drivers/batt_smbus/module.mk b/src/drivers/batt_smbus/module.mk
deleted file mode 100644
index b32ea6e..0000000
```

LorenzMeier commented:
@ksschwabe You should integrate this commit: https://github.com/3drobotics/PX4Firmware-solo/commit/8bd17a4b0cd9c1432cb57a9a80a215692f532370

But I'm not sure how to handle the button press between different manufacturers - a switch or similar based on manufacturer ID would be great.

ksschwabe commented:
@LorenzMeier : I've incorporated the changes, but I have commented out the section that reads the button press. They are reading the "ManufacturerData()" register, which according to the [Smart Battery Data Specification](http://sbs-forum.org/specs/sbdat110.pdf), differs from manufacturer to manufacturer: 
> The ManufacturerData() function may be used to access the manufacturer's data area.  The information and its format are proprietary, but might include items such as: lot codes, number of deep cycles, discharge patterns, deepest discharge, etc.  The Smart Battery manufacturer is free to use this data as they see fit.

Without knowing what battery management chip they are using, I don't know what I need to check to determine that I am communicating with their chip, and hence to know whether to check that register or not.

Any ideas?

LorenzMeier commented:
@liamstask Is there a way for us to tell a Solo battery apart from something else?

LorenzMeier commented:
@Gussy Let us know if you have an idea how to identify Solo's battery so we can properly condition the button code.

Gussy commented:
@LorenzMeier The best way to do this is probably using a "button press supported" lookup table containing ManufacturerName and DeviceName. I don't have values for those fields from Solo batteries on hand though.

liamstask commented:
could possibly add `DeviceChemistry` as well, in case that somehow changes in the future.

ksschwabe commented:
@Gussy @liamstask @LorenzMeier 

Could someone give me the strings for `ManufacturerName`, `DeviceName`, and `DeviceChemisty` for the solo battery so that I can incorporate it? I will probably just have a flag set for the solo battery and if it finds the battery to be the solo battery, then it can check the `ManufacturerData` for the button press.

liamstask commented:
@ksschwabe sure thing, we'll collect that info and get back to you.

LorenzMeier commented:
@liamstask kind reminder on the Solo battery ID stuff.

liamstask commented:
sorry about that - have been syncing with somebody at 3DR who had been working on gathering that info. Hoping to have it available soon.

liamstask commented:
below is prelim output from a single battery. still planning to run this on a broader sample of batteries to verify they respond consistently.

![screenshot](https://cloud.githubusercontent.com/assets/51256/12247591/83c8e39e-b869-11e5-9383-88a2111989f6.png)

liamstask commented:
readings from a few more units:

![battery1](https://cloud.githubusercontent.com/assets/51256/12271736/8108d6e8-b911-11e5-92ab-e20c90b236c0.PNG)
![battery2](https://cloud.githubusercontent.com/assets/51256/12271735/81067b00-b911-11e5-8a2f-d52c53bfabe7.PNG)
![battery3](https://cloud.githubusercontent.com/assets/51256/12271737/8108e67e-b911-11e5-8d8a-99bb2b96f593.PNG)

LorenzMeier commented:
@ksschwabe Given you have now additional input, could you wrap this up?

ksschwabe commented:
@LorenzMeier : I have incorporated the check for if it is a 3DR Solo battery and rebased from master. 

The driver checks if it has the Manufacturer Name, Device Name and Device Chemistry. If not it tries to read it out when cycle() executes. If that data has been updated it performs the check for if the battery is a 3DR Solo battery.

If the Manufacturer Name, Device Name, or Device Chemistry had already been read out at some point, it does not perform the check for if it is a Solo battery.

If the flag for a Solo battery having been detected has been set, it tries to check for the button press, as was done in the 3DR branch of the code previously using the ` ManufacturerData`.

I have tried to check the logic using a Texas Instruments bq40z50 battery management chip. It seems to work, but still needs to be tested on an actual Solo battery before it can be merged in. Perhaps @liamstask can can check if this works on a Solo battery?

I also have a few line comments in the code regarding style. Perhaps you would like to check that and give me some feedback if I should make any changes.

ksschwabe commented on src/drivers/batt_smbus/batt_smbus.cpp:
```diff
@@ -381,11 +590,53 @@ BATT_SMBUS::cycle()
+	if (perform_solo_battry_check) {
+		warnx("Checking solo battery");
+		check_if_solo_battery();
+	}
```
Here I try and read out the Manufacturer Name, Device Name and Device Chemistry. Should this be in cycle(), or should it rather be moved to the start function, since it really has to be only run once?

ksschwabe commented on src/drivers/batt_smbus/batt_smbus.cpp:
```diff
@@ -426,14 +677,42 @@ BATT_SMBUS::cycle()
+					new_report.is_powering_off = false;
+				}
+			}
+		}
```
This check for the button press is the same check that was performed on the 3drobotics/PX4Firmware-solo@8bd17a4. It looks at the `ManufacturerData` and determines if the battery button has been pressed.

ksschwabe commented on src/drivers/batt_smbus/batt_smbus.cpp:
```diff
@@ -77,30 +77,48 @@
+};
+
+// Declaration of the solo battery data, as determined by reading out the data from multiple 3DR Solo batteries
+const struct battery_type solo_battery = {(char *)"BMTPOW", (char *)"MA03", (char *)"LIon"};
```
This defines the characteristics or expected 'properties' for a Solo battery against which the Manufacturer Name, Device Name and Device Chemistry is checked when it is read out from the SMBus .

liamstask commented:
does solo specific functionality belong in this module? this would set the precedent of adding vehicle specific support here, which seems less than ideal.

ksschwabe commented:
@liamstask : 

> does solo specific functionality belong in this module? this would set the precedent of adding vehicle specific support here, which seems less than ideal. perhaps there's a better place to add it?

I tend to agree. But I have absolutely no idea how I should go about splitting that particular part out. Unless you want to turn this batt_smbus class into parent/general class and then have a Solo-specific battery driver that inherits from this class. But then you would need some sort of compile time flag to tell it to run the Solo-specific driver instead of the normal batt_smbus driver. And then if the batt_smbus driver is modified, the Solo-driver might be left unmodified. It makes maintaining the structure a bit messy.

liamstask commented:
my personal preference would be to avoid inheritance in this case, and allow a solo-specific module/layer to simply query the `batt_smbus` module for the info it's interested in.

i'm not sure where the appropriate place to add this functionality is, though. perhaps @julianoes has some insight?

LorenzMeier commented:
@ksschwabe @julianoes Would be nice to settle this and not drop the ball here. Thanks!

ksschwabe commented:
@LorenzMeier: I'm waiting on feedback from @liamstask and @julianoes on how they want the architecture of the driver to change. As far as I can tell (without having been able to test it on an actual Solo battery) it works "as is" for the moment, but may require some architectural changes to better fit the modularity of the Px4 project.

LorenzMeier commented:
@liamstask @julianoes Would be good to settle the discussion before this becomes stale.

liamstask commented:
I'm not sure I have anything further to contribute on this issue - @ksschwabe, which feedback are you waiting for?

My comments are the same as my last one: separating the solo-specific portions from the smbus portions would be good, i don't think the solo-specific portions should be implemented via inheritance from the smbus instance, and I'm not sure where the solo-specific portions should live :D

LorenzMeier commented:
Merging as is - refactoring can always happen later.

LorenzMeier commented:
Rebased and applied, thanks everybody!

ksschwabe commented:
@LorenzMeier: Ok. Let me know if you see any problems.
